### PR TITLE
SF-1854 Fix note thread permissions

### PR DESCRIPTION
**Pull request by Nateowami**
- I've reformatted the permissions to make them more maintainable
- All Paratext roles have been given view permission on the note threads (thus fixing the actual bug).

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1676)

### CODE REVIEW

**Review by RaymondLuong3 — CHANGES_REQUESTED** [review]
I think the layout of the previous version of the SF projects rights was more readable than having a tuple ProjectRights type. But this new type also has its advantages of being more compact. I think in the end the compactness will be the bigger pay-off.

> Reviewed 1 of 3 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-:-NMembsn5Z_5zS4-WSJf:b-6fds7q)* status: all files reviewed, 5 unresolved discussions (waiting on @Nateowami)

___
*[`src/RealtimeServer/common/models/project-rights.ts` line 15 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-NMeNEOW7qyBpL_DawTE:-NMejYII8KpSZxKZ0OAa:b-nclwmo) ([raw file](https://github.com/sillsdev/web-xforge/blob/95589bd47d343698f42d6ecd4774a2737e122cff/src/RealtimeServer/common/models/project-rights.ts#L15)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

This is basically a tuple. I debated whether to do this or not. I also tried to change `string` to something more specific, but since the domains are specified for SFProject (not Project), it causes all sorts of challenges to make the domains type safe (you basically have to make a whole bunch of things generic and take a type parameter, and it seemed like more trouble than it was worth).

I replaced `Operation` because it's an enum, and when an enum value is expected, you can't provide a string, such as `'edit'`, even if that's one of the values of the `Operation` enum. I replaced it with the union of all `Operation` values, so now it expects string values.
</blockquote></details>

Oh boy, a comment on the type would help since it is not intuitive about the enum thing.
___
*[`src/RealtimeServer/scriptureforge/models/sf-project-rights.ts` line 18 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-NMekGic88uCmxbcXvAt:-NMekGic88uCmxbcXvAu:b2sy3wm) ([raw file](https://github.com/sillsdev/web-xforge/blob/3676f424c1f2fc37b42836cc87373de28a3aaf00/src/RealtimeServer/scriptureforge/models/sf-project-rights.ts#L18)):*
> ```XML
> }
> 
> const rightsByRole: Record<SFProjectRole, { [domain in `${SFProjectDomain}`]?: `${Operation}`[] }> = {
> ```

I would leave a comment stating why the SFProjectDomain is optional. Just like what you typed above.



_Code quote:_
```XML
{ [domain in `${SFProjectDomain}`]?:
```
___
*[`src/RealtimeServer/scriptureforge/models/sf-project-rights.ts` line 83 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-NMekvt84502-ujUhA2T:-NMekvt84502-ujUhA2U:bw4v7m6) ([raw file](https://github.com/sillsdev/web-xforge/blob/3676f424c1f2fc37b42836cc87373de28a3aaf00/src/RealtimeServer/scriptureforge/models/sf-project-rights.ts#L83)):*
> ```XML
>     texts: ['view', 'edit'],
>     questions: ['view', 'create', 'edit', 'delete'],
>     answers: ['view', 'delete', 'edit_own'],
> ```

Don't you need to give admins permission to create an answer here?



_Code quote:_
```XML
    answers: ['view', 'delete', 'edit_own'],
```
___
*[`src/RealtimeServer/scriptureforge/models/sf-project-rights.ts` line 125 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-NMelvf_C7dSVNeIFOYG:-NMelvf_C7dSVNeIFOYH:bkqq1ur) ([raw file](https://github.com/sillsdev/web-xforge/blob/3676f424c1f2fc37b42836cc87373de28a3aaf00/src/RealtimeServer/scriptureforge/models/sf-project-rights.ts#L125)):*
> ```XML
> 
>     for (const [role, rights] of Object.entries(rightsByRole)) {
>       const rightsForRole: ProjectRight[] = [];
> ```

Hmm, this triple nested for loop makes me think that the data is structures in an inconvenient way. But it is also structured in a logical way, with each domain having a set of operations. So this should be ok.



_Code quote:_
```XML
    for (const [role, rights] of Object.entries(rightsByRole)) {
      const rightsForRole: ProjectRight[] = [];
      for (const [domain, operations] of Object.entries(rights)) {
        for (const operation of operations) rightsForRole.push([domain, operation]);
      }
      this.addRights(role, rightsForRole);
```

File: src/RealtimeServer/common/models/project-rights.ts [hunk 1]
```diff
@@ -12,10 +12,7 @@ export enum Operation {
   ViewOwn = 'view_own'
 }
 
-export interface ProjectRight {
-  projectDomain: string;
-  operation: Operation;
-}
+export type ProjectRight = [string, `${Operation}`];
```
Comment (pmachapman) [inline src/RealtimeServer/common/models/project-rights.ts:15]: Should we label the tuple elements? i.e. 
```typescript
export type ProjectRight = [projectDomain: string, operation: `${Operation}`];
```
See https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/#labeled-tuple-elements

**Review by pmachapman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1676)* status: all files reviewed, 5 unresolved discussions (waiting on @Nateowami and @RaymondLuong3)

___
*[`src/RealtimeServer/common/models/project-rights.ts` line 15 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-NMeNEOW7qyBpL_DawTE:-NMeoueH3nO6iOHmLe2j:bi6t7n) ([raw file](https://github.com/sillsdev/web-xforge/blob/95589bd47d343698f42d6ecd4774a2737e122cff/src/RealtimeServer/common/models/project-rights.ts#L15)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Oh boy, a comment on the type would help since it is not intuitive about the enum thing.
</blockquote></details>

Should we label the tuple elements? i.e. 
```typescript
export type ProjectRight = [projectDomain: string, operation: `${Operation}`];
```
See https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/#labeled-tuple-elements

**Review by Nateowami — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1676)* status: 1 of 3 files reviewed, 5 unresolved discussions (waiting on @pmachapman and @RaymondLuong3)

___
*[`src/RealtimeServer/common/models/project-rights.ts` line 15 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-NMeNEOW7qyBpL_DawTE:-NMfPoRmCPIc6AH_-CER:b-6rns8v) ([raw file](https://github.com/sillsdev/web-xforge/blob/95589bd47d343698f42d6ecd4774a2737e122cff/src/RealtimeServer/common/models/project-rights.ts#L15)):*
<details><summary><i>Previously, pmachapman (Peter Chapman) wrote…</i></summary><blockquote>

Should we label the tuple elements? i.e. 
```typescript
export type ProjectRight = [projectDomain: string, operation: `${Operation}`];
```
See https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/#labeled-tuple-elements
</blockquote></details>

Wasn't aware of the ability to label tuples. Done.
___
*[`src/RealtimeServer/scriptureforge/models/sf-project-rights.ts` line 18 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-NMekGic88uCmxbcXvAt:-NMfRHCC5Qcy0CwXVjJY:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/3676f424c1f2fc37b42836cc87373de28a3aaf00/src/RealtimeServer/scriptureforge/models/sf-project-rights.ts#L18)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

I would leave a comment stating why the SFProjectDomain is optional. Just like what you typed above.
</blockquote></details>

Done.
___
*[`src/RealtimeServer/scriptureforge/models/sf-project-rights.ts` line 83 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-NMekvt84502-ujUhA2T:-NMfHkvQ64V-qZUUoeGH:briro7x) ([raw file](https://github.com/sillsdev/web-xforge/blob/3676f424c1f2fc37b42836cc87373de28a3aaf00/src/RealtimeServer/scriptureforge/models/sf-project-rights.ts#L83)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Don't you need to give admins permission to create an answer here?
</blockquote></details>

No, I specifically avoided making that change in this PR because it breaks a number of tests. I was planning to do that as a followup after we fix the existing permission issue.

Though I did need to give the "edit_own" permission since we (bizarrely) have a test that admins can edit their own answers, even though they can't create them.

**Review by RaymondLuong3 — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-:-NN7nWVy6M3uyYXMcpf5:bg7igkv)* status: all files reviewed, 2 unresolved discussions (waiting on @Nateowami)

___
*[`src/RealtimeServer/scriptureforge/models/sf-project-rights.ts` line 83 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1676#-NMekvt84502-ujUhA2T:-NN7nT0gCM5cVK1ITkVn:b6wyrya) ([raw file](https://github.com/sillsdev/web-xforge/blob/3676f424c1f2fc37b42836cc87373de28a3aaf00/src/RealtimeServer/scriptureforge/models/sf-project-rights.ts#L83)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

No, I specifically avoided making that change in this PR because it breaks a number of tests. I was planning to do that as a followup after we fix the existing permission issue.

Though I did need to give the "edit_own" permission since we (bizarrely) have a test that admins can edit their own answers, even though they can't create them.
</blockquote></details>

Nice, good thought.

**Comment by marksvc** [comment]
Nathaniel asked me to merge this thru.

**Pull request merged**